sycl/test-e2e/Matrix/joint_matrix_activation_impl.hpp

```diff
@@ -0,0 +1,120 @@
+//==-------- joint_matrix_down_convert_impl.hpp  - DPC++ joint_matrix-------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include <iostream>
+
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental::matrix;
+
+constexpr size_t TM = 8;
+// TN and TK must be the same for this test.
+constexpr size_t TN = 16;
+constexpr size_t TK = 16;
+
+template <Activation act, size_t TM, size_t TK, typename Group, typename Tsrc,
+          typename Tdest, use UseSrc, use UseDest, layout LayoutSrc,
+          layout LayoutDest>
+void applyActivation(
+    Group &sg, joint_matrix<Group, Tsrc, UseSrc, TM, TK, LayoutSrc> &sub_c,
+    joint_matrix<Group, Tdest, UseDest, TM, TN, LayoutDest> &sub_a) {
+  if constexpr (act == Activation::None) {
+    joint_matrix_copy(sg, sub_c, sub_a);
+  } else if constexpr (act == Activation::ReLU) {
+
+    joint_matrix_apply(
+        sg, sub_c, [=](float &x) { x = sycl::max(static_cast<float>(0), x); });
+    joint_matrix_copy(sg, sub_c, sub_a);
+
+  } else if constexpr (act == Activation::Sigmoid) {
+    joint_matrix_apply(sg, sub_c,
+                       [=](float &x) { x = 1.0f / (1.0f + sycl::exp(-x)); });
+    joint_matrix_copy(sg, sub_c, sub_a);
+  }
+  return;
+}
```

Comment:
what is the purpose of having different activation functions in the test? Is all that necessary to reproduce the bug?
Can it be simplified?

Reply:
Reproduction of the bug is simpler and is the one I attached to the jira.
However, I added these activations usage in the test to imitate more the real workload this bug came from.
Users often complained about the simplifications of these tests.
While simple tests are what we also need, having some close-to-real usage will be also valuable to users.

Reply:
By LLVM testing guide, "a regression test containing just enough code to reproduce the problem should be written" (https://llvm.org/docs/TestingGuide.html).
SYCL E2E regression tests suite is not the right place for "more real workload", in my opinion.
We have a separate repository for oneAPI samples here: https://github.com/oneapi-src/oneAPI-samples/tree/main/DirectProgramming/C%2B%2BSYCL
SYCL E2E tests should be as simple as possible to be easy to maintain and reproduce problems and fast.

Reply:
That is not true. For instance, https://github.com/intel/llvm/tree/sycl/sycl/test-e2e/ESIMD has a lot of real benchmarks like histogram, stencil, kmeans.
This test I am adding can be categorized as microbenchmark. It triggers a bug and also gives more usage of the apply function as part of a function call. It does not add much to the reproducer other than extra couple of arithmetic operations. 

>>> It also add this new outlining functionality which also impacts the SYCL compiler.
Here, I meant that adding `__always_inline` at the code level made the bug go away. So if the SYCL compiler outlined the function, we would not have seen the bug.

Reply:
Not sure, what exactly is not true. I don't know the reasons behind real benchmarks in ESIMD E2E tests, but it doesn't look right. ESIMD SYCL E2E tests structure should not be considered as a reference.

The test that you are adding should go to IGC repo (of course not as-is). The bug is in IGC and is triggered by different variant of IR.
"__always_inline" makes the bug to be not exposed, but it doesn't mean it is bug in SYCL compiler.
So, I don't see, why this test should be in SYCL E2E at all.

Reply:
I think we got off topic, the new test I am adding is needed as a reproducer for the regression which is legitimate to add here. 
What it adds over a simpler reproducer is the sigmoid operation. 
The `sigmoid` operation is also important and legitimate to add as it tests new operations with the apply that we don't test before like `exp`.

Reply:
No, I don't think we are off topic. My point is the bug was in IGC, and regression test for this bug should be added to IGC repo, not SYCL compiler repo.